### PR TITLE
Improve error response handling

### PR DESCRIPTION
**Pull request by MykolaVasyk**
# Title

Enhancing error response handling

## Motivation

Error handling is a critical aspect of any application's functionality. We wanted to our system's error response handling to effectively describe issues to users. So it was necessary to improve our error response mechanism to enhance user experience.

## Modifications

Standardized Error Codes: Introduced a generalized and standardized set of error codes to categorize different types of errors for better identification and resolution.
Detailed Error Messages: Enhanced error messages to provide more detailed and actionable information, including relevant context, possible causes, and suggested actions for resolution.
Added tests: Error handling functionality is covered by tests.

## Result

Improved user experience: Users will receive more informative and user-friendly error messages.
Streamlined troubleshooting: Developers and support teams will have access to detailed error logs and standardized error codes, enabling quicker identification and resolution of issues.
Enhanced system reliability with better error handling mechanisms in place and testings.

### CODE REVIEW

File: Sources/AppStoreConnectClient/AppStoreConnectClient.swift [hunk 1]
```diff
@@ -38,7 +38,7 @@ public struct AppStoreConnectClient {
     /// Initializes the client with App Store Connect credentials.
     /// - Parameter credentials: Credentials needed to authenticate with the App Store Connect API.
     /// - Throws: An error if initialization fails.
-    public init(with credentials: Credentials) throws {
+    public init(with credentials: Credentials) async throws {
```
Comment (mcmikius) [inline Sources/AppStoreConnectClient/AppStoreConnectClient.swift]: Why it async?

File: Sources/AppStoreConnectClient/AppStoreConnectClient.swift [hunk 2]
```diff
@@ -60,40 +60,92 @@ public struct AppStoreConnectClient {
...
+            }
         case .undocumented(let statusCode, _):
             throw AppStoreConnectError.serverError(errorCode: statusCode)
         }
     }
     
+    /// Converts an error response into a formatted string.
+    /// - Parameter response: The error response object.
+    /// - Returns: A formatted string describing the error response, or `nil` if the response is empty.
+    private func handleError(from response: Components.Schemas.ErrorResponse) -> String? {
```
Comment (mcmikius) [inline Sources/AppStoreConnectClient/AppStoreConnectClient.swift]: Move it to error implementation

File: Sources/AppStoreConnectClient/AppStoreConnectClient.swift [hunk 3]
```diff
@@ -60,40 +60,92 @@ public struct AppStoreConnectClient {
...
+        default:
+            return try await errorResponse(from: response)
+        }
+        throw AppStoreConnectError.unexpectedError(errors: "\(response)")
+    }
+    
+    /// Handles error responses returned by the App Store Connect API.
+    /// - Parameter response: The response received from the API.
+    /// - Throws: An error of type `AppStoreConnectError` if the response indicates an error.
+    func errorResponse(
```
Comment (mcmikius) [inline Sources/AppStoreConnectClient/AppStoreConnectClient.swift]: Maybe it private

File: Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift [hunk 4]
```diff
@@ -37,41 +37,65 @@ final class AppStoreConnectClientTests: XCTestCase {
...
+        let text = """
+            \nThe request failed with: \(statusCode), BAD_REQUEST, Invalid Input, The provided data failed validation.
+            """
+        let expectedError = AppStoreConnectError.badRequest(errors: text)
         do {
             _ = try await client.fetchApps()
             XCTFail("Expected error not thrown")
         } catch {
-            XCTAssertEqual(AppStoreConnectError.badRequest.localizedDescription, error.localizedDescription)
+            if let appStoreConnectError = error as? AppStoreConnectError {
```
Comment (mcmikius) [inline Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift]: Delete ifelse and improve

File: Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift [hunk 5]
```diff
@@ -37,41 +37,65 @@ final class AppStoreConnectClientTests: XCTestCase {
...
+        let text = """
+            \nThe request failed with: \(statusCode), FORBIDDEN, Access Denied, You do not have permission to access this resource.
+            """
+        let expectedError = AppStoreConnectError.forbidden(errors: text)
         do {
             _ = try await client.fetchApps()
             XCTFail("Expected error not thrown")
         } catch {
-            XCTAssertEqual(AppStoreConnectError.forbidden.localizedDescription, error.localizedDescription)
+            if let appStoreConnectError = error as? AppStoreConnectError {
```
Comment (mcmikius) [inline Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift]: Delete ifelse and improve

File: Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift [hunk 6]
```diff
@@ -37,41 +37,65 @@ final class AppStoreConnectClientTests: XCTestCase {
...
+        let text = """
+            \nThe request failed with: \(statusCode), UNAUTHORIZED, Unauthorized, Authentication credentials were missing or incorrect.
+            """
+        let expectedError = AppStoreConnectError.unauthorized(errors: text)
         do {
             _ = try await client.fetchApps()
             XCTFail("Expected error not thrown")
         } catch {
-            XCTAssertEqual(AppStoreConnectError.unauthorized.localizedDescription, error.localizedDescription)
+            if let appStoreConnectError = error as? AppStoreConnectError {
```
Comment (mcmikius) [inline Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift]: Delete ifelse and improve

File: Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift [hunk 7]
```diff
@@ -109,11 +138,20 @@ final class AppStoreConnectClientTests: XCTestCase {
             _ = try await client.fetchVersions(for: app)
             XCTFail("Expected error not thrown")
         } catch {
-            XCTAssertEqual(AppStoreConnectError.badRequest.localizedDescription, error.localizedDescription)
+            if let appStoreConnectError = error as? AppStoreConnectError {
```
Comment (mcmikius) [inline Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift]: Delete ifelse and improve

File: Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift [hunk 8]
```diff
@@ -122,11 +160,20 @@ final class AppStoreConnectClientTests: XCTestCase {
             _ = try await client.fetchVersions(for: app)
             XCTFail("Expected error not thrown")
         } catch {
-            XCTAssertEqual(AppStoreConnectError.forbidden.localizedDescription, error.localizedDescription)
+            if let appStoreConnectError = error as? AppStoreConnectError {
```
Comment (mcmikius) [inline Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift]: Delete ifelse and improve

File: Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift [hunk 9]
```diff
@@ -135,11 +182,20 @@ final class AppStoreConnectClientTests: XCTestCase {
             _ = try await client.fetchVersions(for: app)
             XCTFail("Expected error not thrown")
         } catch {
-            XCTAssertEqual(AppStoreConnectError.notFound.localizedDescription, error.localizedDescription)
+            if let appStoreConnectError = error as? AppStoreConnectError {
```
Comment (mcmikius) [inline Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift]: Delete ifelse and improve

File: Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift [hunk 10]
```diff
@@ -148,7 +204,11 @@ final class AppStoreConnectClientTests: XCTestCase {
             _ = try await client.fetchVersions(for: app)
             XCTFail("Expected error not thrown")
         } catch {
-            XCTAssertEqual(AppStoreConnectError.unauthorized.localizedDescription, error.localizedDescription)
+            if let appStoreConnectError = error as? AppStoreConnectError {
```
Comment (mcmikius) [inline Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift]: Delete ifelse and improve

**Review by mcmikius — COMMENTED** [review]
Need resolve comments

File: Sources/AppStoreConnectClient/AppStoreConnectClient.swift (hunk 1 above)
Comment (MykolaVasyk) [reply Sources/AppStoreConnectClient/AppStoreConnectClient.swift]: Deleted.

File: Sources/AppStoreConnectClient/AppStoreConnectClient.swift (hunk 3 above)
Comment (MykolaVasyk) [reply Sources/AppStoreConnectClient/AppStoreConnectClient.swift]: Added private.

File: Sources/AppStoreConnectClient/AppStoreConnectClient.swift (hunk 2 above)
Comment (AndriyVasyk) [reply Sources/AppStoreConnectClient/AppStoreConnectClient.swift]: Need to correct the name of this function - it should be something like "errorString(from..." or "errorDescription(from..."

Changed to errorDescription.

File: Sources/AppStoreConnectClient/AppStoreConnectClient.swift [hunk 11]
```diff
@@ -60,40 +60,92 @@ public struct AppStoreConnectClient {
                 return json.data.compactMap({ Application(schema: $0) })
             }
         case .badRequest(let result):
-            throw AppStoreConnectError.badRequest
+            switch result.body {
```
Comment (AndriyVasyk) [inline Sources/AppStoreConnectClient/AppStoreConnectClient.swift]: I believe that we can use 'result.body' as param for 'handleError'. In this case we will be able to omit  duplications of switch/cases

File: Sources/AppStoreConnectClient/AppStoreConnectClient.swift [hunk 12]
```diff
@@ -60,40 +60,92 @@ public struct AppStoreConnectClient {
...
+        }
+        throw AppStoreConnectError.unexpectedError(errors: "\(response)")
+    }
+    
+    /// Handles error responses returned by the App Store Connect API.
+    /// - Parameter response: The response received from the API.
+    /// - Throws: An error of type `AppStoreConnectError` if the response indicates an error.
+    func errorResponse(
+        from response: Operations.apps_hyphen_appStoreVersions_hyphen_get_to_many_related.Output
+    ) async throws -> [Release] {
```
Comment (AndriyVasyk) [inline Sources/AppStoreConnectClient/AppStoreConnectClient.swift]: Do not see any reason why this function should be async

File: Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift [hunk 13]
```diff
@@ -27,7 +27,7 @@ import XCTest
 
 final class AppStoreConnectClientTests: XCTestCase {
     func testFetchAppsSuccess() async throws {
-        var mockClient = MockAPIClient()
+        var mockClient = MockAPIClient(result: .ok)
```
Comment (AndriyVasyk) [inline Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift]: Why we need this change? Looks like the next line (mockClient.result = .ok) does the same

File: Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift [hunk 14]
```diff
@@ -37,41 +37,65 @@ final class AppStoreConnectClientTests: XCTestCase {
     }
     
     func testFetchAppsBadRequest() async throws {
+        let statusCode = 400
         var mockClient = MockAPIClient()
         mockClient.result = .badRequest
         let client = AppStoreConnectClient(client: mockClient)
-        
+        let text = """
+            \nThe request failed with: \(statusCode), BAD_REQUEST, Invalid Input, The provided data failed validation.
```
Comment (AndriyVasyk) [inline Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift]: We either need to add this format string to some Contacts or avoid comparing errors (instead just catch particular error in separate catch block)

File: Tests/AppStoreConnectClientTests/TestHelpers/MockObjects.swift [hunk 15]
```diff
@@ -26,13 +27,82 @@ import OpenAPIRuntime
...
+    static var errorResponseArray: Components.Schemas.ErrorResponse = {
+        let payload1 = Components.Schemas.ErrorResponse.errorsPayloadPayload(
+            status: "400",
+            code: "BED_REQUEST",
+            title: "Server Error",
+            detail: "An unexpected error"
+        )
+        let payload2 = Components.Schemas.ErrorResponse.errorsPayloadPayload(
+            status: "500",
+            code: "INTERNAL_SERVER_ERROR",
```
Comment (AndriyVasyk) [inline Tests/AppStoreConnectClientTests/TestHelpers/MockObjects.swift]: Need some comment regarding the source of these codes (INTERNAL_SERVER_ERROR)

File: Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift (hunk 4 above)
Comment (MykolaVasyk) [reply Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift]: Changed.

File: Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift (hunk 5 above)
Comment (MykolaVasyk) [reply Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift]: Changed.

File: Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift (hunk 6 above)
Comment (MykolaVasyk) [reply Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift]: Changed.

File: Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift (hunk 7 above)
Comment (MykolaVasyk) [reply Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift]: Changed.

File: Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift (hunk 8 above)
Comment (MykolaVasyk) [reply Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift]: Changed.

File: Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift (hunk 9 above)
Comment (MykolaVasyk) [reply Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift]: Changed.

File: Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift (hunk 10 above)
Comment (MykolaVasyk) [reply Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift]: Changed.

File: Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift (hunk 13 above)
Comment (MykolaVasyk) [reply Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift]: Changed.

File: Sources/AppStoreConnectClient/AppStoreConnectClient.swift (hunk 11 above)
Comment (MykolaVasyk) [reply Sources/AppStoreConnectClient/AppStoreConnectClient.swift]: Changed.

File: Sources/AppStoreConnectClient/AppStoreConnectClient.swift (hunk 12 above)
Comment (MykolaVasyk) [reply Sources/AppStoreConnectClient/AppStoreConnectClient.swift]: Changed.

File: Tests/AppStoreConnectClientTests/TestHelpers/MockObjects.swift (hunk 15 above)
Comment (MykolaVasyk) [reply Tests/AppStoreConnectClientTests/TestHelpers/MockObjects.swift]: Changed.

File: Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift (hunk 14 above)
Comment (MykolaVasyk) [reply Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift]: Changed.

File: Sources/AppStoreConnectClient/AppStoreConnectClient.swift (hunk 2 above)
Comment (MykolaVasyk) [reply Sources/AppStoreConnectClient/AppStoreConnectClient.swift]: moved to AppStoreConnectError.swift

File: Sources/AppStoreConnectClient/AppStoreConnectClient.swift [hunk 16]
```diff
@@ -55,45 +55,73 @@ public struct AppStoreConnectClient {
...
+        }
+        throw AppStoreConnectError.unexpectedError(errors: "\(response)")
+    }
+    
+    /// Handles error responses returned by the App Store Connect API.
+    /// - Parameter response: The response received from the API.
+    /// - Throws: An error of type `AppStoreConnectError` if the response indicates an error.
+    private func errorResponse(
+        from response: Operations.apps_hyphen_appStoreVersions_hyphen_get_to_many_related.Output
+    ) throws -> [Release] {
```
Comment (AndriyVasyk) [inline Sources/AppStoreConnectClient/AppStoreConnectClient.swift]: do we need to return anything from this func?

File: Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift [hunk 17]
```diff
@@ -97,7 +101,7 @@ final class AppStoreConnectClientTests: XCTestCase {
         let apps = try await client.fetchVersions(for: app)
         XCTAssertEqual(1, apps.count)
         XCTAssertEqual(apps.first?.appStoreState, "ACCEPTED")
-        XCTAssertEqual(apps.first?.version, "1.1.1")
+        XCTAssertEqual(apps.first?.version, "Foo.Bar.Baz")
```
Comment (AndriyVasyk) [inline Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift]: need to use corresponding mock object here

File: Sources/AppStoreConnectClient/AppStoreConnectClient.swift [hunk 18]
```diff
@@ -55,45 +55,73 @@ public struct AppStoreConnectClient {
...
         switch response {
         case .ok(let okResponse):
-            switch okResponse.body {
-            case .json(let json):
-                return json.data.compactMap({ Application(schema: $0) })
-            }
+            return try okResponse.body.json.data.compactMap({ Application(schema: $0) })
         case .badRequest(let result):
-            throw AppStoreConnectError.badRequest
+            throw AppStoreConnectError.badRequest(
```
Comment (AndriyVasyk) [inline Sources/AppStoreConnectClient/AppStoreConnectClient.swift]: i believe that we can reuse "errorResponse" here too

File: Sources/AppStoreConnectClient/AppStoreConnectClient.swift (hunk 16 above)
Comment (MykolaVasyk) [reply Sources/AppStoreConnectClient/AppStoreConnectClient.swift]: Deleted "return".

Comment (mcmikius) [reply Sources/AppStoreConnectClient/AppStoreConnectClient.swift]: this function does not return anything, so you need to remove what it returns - [release]

Comment (mcmikius) [reply Sources/AppStoreConnectClient/AppStoreConnectClient.swift]: I think function name is incorrect, cause this function is handling error response. Maybe correct name is handleErrors( or something like that.

File: Sources/AppStoreConnectClient/AppStoreConnectError.swift [hunk 19]
```diff
@@ -29,13 +29,31 @@ public enum AppStoreConnectError: Error, Equatable {
...
     /// The requested resource was not found.
-    case notFound
-    /// The credentials is empty.
-    case noHaveCredentials
+    case notFound(errors: String?)
+    /// An unexpected error occurred.
+    case unexpectedError(errors: String?)
+}
+
+extension AppStoreConnectError {
```
Comment (mcmikius) [inline Sources/AppStoreConnectClient/AppStoreConnectError.swift]: Move to a separate file AppStoreConnectError+Extensions

File: Sources/AppStoreConnectClient/AppStoreConnectError.swift [hunk 20]
```diff
@@ -29,13 +29,31 @@ public enum AppStoreConnectError: Error, Equatable {
...
+    /// Converts an error response into a formatted string.
+    /// - Parameter response: The error response object.
+    /// - Returns: A formatted string describing the error response, or `nil` if the response is empty.
+    static func errorDescription(from response: Components.Schemas.ErrorResponse) -> String? {
+        var stringError = ""
+        response.errors?.forEach { error in
+            stringError.append("""
+                \nThe request failed with: \(error.status), \(error.code), \(error.title), \(error.detail).
```
Comment (mcmikius) [inline Sources/AppStoreConnectClient/AppStoreConnectError.swift]: Failed with

**Review by mcmikius — COMMENTED** [review]
Need resolve new comments

File: Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift (hunk 17 above)
Comment (MykolaVasyk) [reply Tests/AppStoreConnectClientTests/AppStoreConnectClientTests.swift]: Fixed.

File: Sources/AppStoreConnectClient/AppStoreConnectClient.swift (hunk 18 above)
Comment (MykolaVasyk) [reply Sources/AppStoreConnectClient/AppStoreConnectClient.swift]: Refactored.

File: Sources/AppStoreConnectClient/AppStoreConnectError.swift (hunk 19 above)
Comment (MykolaVasyk) [reply Sources/AppStoreConnectClient/AppStoreConnectError.swift]: Moved to separate file.

File: Sources/AppStoreConnectClient/AppStoreConnectError.swift (hunk 20 above)
Comment (MykolaVasyk) [reply Sources/AppStoreConnectClient/AppStoreConnectError.swift]: Changed text.

File: Sources/AppStoreConnectClient/AppStoreConnectClient.swift [hunk 21]
```diff
@@ -53,47 +53,92 @@ public struct AppStoreConnectClient {
...
+            let error = try error(from: response)
+            throw error
+        }
+        return result
+    }
+    
+    /// Handles error responses returned by the App Store Connect API when fetching app store versions.
+    /// - Parameter response: The response received from the API.
+    /// - Throws: An error of type `AppStoreConnectError` if the response indicates an error.
+    private func error(
```
Comment (mcmikius) [inline Sources/AppStoreConnectClient/AppStoreConnectClient.swift]: Func handleErrors(from ...

File: Sources/AppStoreConnectClient/AppStoreConnectClient.swift [hunk 22]
```diff
@@ -53,47 +53,92 @@ public struct AppStoreConnectClient {
...
+            
+        default:
+            return AppStoreConnectError.unexpectedError(errors: "\(response)")
+        }
+    }
+    
+    /// Handles error responses returned by the App Store Connect API when fetching app collections.
+    /// - Parameter response: The response received from the API.
+    /// - Throws: An error of type `AppStoreConnectError` if the response indicates an error.
+    private func error(
```
Comment (mcmikius) [inline Sources/AppStoreConnectClient/AppStoreConnectClient.swift]: Func handleErrors(from ...

File: Sources/AppStoreConnectClient/AppStoreConnectError+Extensions.swift [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+//  Created by Mykola Vasyk on 10.05.2024.
+//
+
+import Foundation
+
+extension AppStoreConnectError {
+    /// Converts an error response into a formatted string.
+    /// - Parameter response: The error response object.
+    /// - Returns: A formatted string describing the error response, or `nil` if the response is empty.
+    static func errorDescription(from response: Components.Schemas.ErrorResponse) -> String? {
```
Comment (mcmikius) [inline Sources/AppStoreConnectClient/AppStoreConnectError+Extensions.swift]: parseErrorDescription(from:) або handleError(from:)

**Review by mcmikius — COMMENTED** [review]
Add minor comments

File: Sources/AppStoreConnectClient/AppStoreConnectError+Extensions.swift (hunk 23 above)
Comment (MykolaVasyk) [reply Sources/AppStoreConnectClient/AppStoreConnectError+Extensions.swift]: Changed to parseErrorDescription.

File: Sources/AppStoreConnectClient/AppStoreConnectClient.swift (hunk 21 above)
Comment (MykolaVasyk) [reply Sources/AppStoreConnectClient/AppStoreConnectClient.swift]: Changed.

File: Sources/AppStoreConnectClient/AppStoreConnectClient.swift (hunk 22 above)
Comment (MykolaVasyk) [reply Sources/AppStoreConnectClient/AppStoreConnectClient.swift]: Changed.

**Review by mcmikius — APPROVED** [review]

**Pull request merged**